### PR TITLE
Tickets/dm 28415

### PR DESCRIPTION
**Pull request by bechtol**
This PR updates the datasets used for unit tests. 

It seems likely that we will need to replace the test datasets often as we need new columns for metric computations, respond to schema changes, change dataset type names, and/or add new metrics. This PR adds a notebook, `tests/data/create_test_data.ipynb`, to show the step-by-step process of creating the test dataset, and automate most of the steps. I'm not sure that this notebook belongs in the `tests/data` directory. We will need to update the README in that directory as well since it is now largely superseded by the notebook.

Note that when generating the test datasets, we use a subset of ci_hsc to keep the file sizes as small as possible. I don't think it is possible to go too much smaller due to requirements on the minimum number of matched observations. 

There are still a number of steps that need to be done by hand, such as updating the expected values for several of the unit tests, e.g., `tests/test_tex_util.py`. 

Currently the yaml files containing expected values are given names that include the metric name. If the metric name changes, the file names will change. Currently the matched catalogs have names that include the dataset type name. If the dataset type name changes, the file names will change. In addition, if moving a metric from patch-scale to tract-scale, the current naming conventions for files would lead to many file name changes. I'm not sure how to best manage those changes. Perhaps we have a data directory that is easy to wholesale replace with newly generated data and expected value files.

I have modified the SConscript to now perform all the KPM tests. Skipping the stellar locus test for now until we resolve the dust maps issue.

### CODE REVIEW

File: tests/data/AB1_design_expected_0_70_i.yaml [hunk 1]
```diff
@@ -2,10 +2,10 @@
...
-  identifier: 44ebc05ab85944c7ac85d4b928d22b81
+  identifier: 2faaa217f6724d1aa31d2abdc44f086d
   name: AB1_design
-identifier: 4a2e6fb41cd1486d8a94dc105e41b7a1
+identifier: ce18ea93b21b489a8a0b1a3d5e670e43
 metric: AB1_design
 notes: {}
-unit: marcsec
-value: 7.662377260108258
+unit: marcsec2
```
Comment (jeffcarlin) [inline tests/data/AB1_design_expected_0_70_i.yaml:10]: This looks like it has units of marcsec^2 -- something is fishy (but not due to anything in this PR, methinks).

File: tests/data/PA1_expected_0_i.yaml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+!<lsst.verify.Measurement>
```
Comment (jeffcarlin) [inline tests/data/PA1_expected_0_i.yaml:1]: Should this (and the other photometry metric files) be PA1_expected_0_70_*.yaml? (I'm curious why it's inconsistent with the others.)

Comment (jeffcarlin) [reply tests/data/PA1_expected_0_i.yaml:1]: Oh, never mind - I see now that you persisted tract-level catalogs along with the patch ones.

File: tests/test_astromKPM.py [hunk 3]
```diff
@@ -86,6 +86,8 @@ def test_af1(self):
         for band in ('i', 'r'):
             catalog, expected = self.load_data(('AF1', band))
             result = task.run(catalog, 'AF1_design')
+            print(result.measurement)
```
Comment (jeffcarlin) [inline tests/test_astromKPM.py:89]: Probably want to remove these print statements before merging.

File: tests/data/shrink_cat.py [hunk 4]
```diff
@@ -8,17 +8,24 @@
...
-    except ValueError as e:
-        print(e)
+    cat.writeFits(outfile)
 
-cat.writeFits(sys.argv[2])
+
+if __name__ == "__main__":
+    infile = sys.argv[1]
+    outfile = sys.argv[2]
+    shrinkCat(infile, outfile)
```
Comment (jeffcarlin) [inline tests/data/shrink_cat.py:31]: Is this a flake8 error I see? (Reminder to run `scons`.) :)

File: tests/test_phot_util.py [hunk 5]
```diff
@@ -64,31 +63,30 @@ def test_calcPhotRepeat(self):
...
         # Ensure measurements are deterministic
-        random.seed(8675309)
+        seed = 8675309
 
-        expected = 72.21113484367118 * u.mmag
+        expected = 72.79576602821936 * u.mmag
         matches, magKey = self.load_data()
-        result = calcPhotRepeat(matches, magKey)
+        result = calcPhotRepeat(matches, magKey, randomSeed=seed)
+        print('LOOK HERE', result['repeatability'])
```
Comment (jeffcarlin) [inline tests/test_phot_util.py:71]: Remove this debug statement?

File: tests/test_phot_util.py [hunk 6]
```diff
@@ -98,21 +96,23 @@ def test_computeWidths(self):
     def test_getRandomDiffRmsInMmags(self):
         """Test random sampling of magnitude diffs."""
         # Ensure measurements are deterministic
-        random.seed(8675309)
+        seed = 8675309
+        rng = np.random.default_rng(seed)
 
-        expected = 848.5281374238564
+        expected = -1237.436867076458
```
Comment (jeffcarlin) [inline tests/test_phot_util.py:102]: I'm disturbed that what you get when you call 867-5309 is an RMS diff of >1 mag, but if that's what comes out, then OK.

**Review by jeffcarlin — APPROVED** [review]
Looks good in general. I didn't test running it, but I trust that you had to do that in order to "test the tests." 

On Monday I'll look into the AB1 units issue.

File: tests/test_phot_util.py (hunk 5 above)
Comment (bechtol) [reply tests/test_phot_util.py:71]: Thanks for catching.

File: tests/test_astromKPM.py (hunk 3 above)
Comment (bechtol) [reply tests/test_astromKPM.py:89]: Thanks. Indeed I had meant to remove these before merging. Thanks for catching.

File: tests/data/shrink_cat.py (hunk 4 above)
Comment (bechtol) [reply tests/data/shrink_cat.py:31]: Thanks. I did run `scons` but didn't see this among the output. I run `flake8` separately now and verify no more of flake8 violations.

File: tests/test_phot_util.py (hunk 5 above)
Comment (bechtol) [reply tests/test_phot_util.py:71]: Removed

File: tests/test_phot_util.py (hunk 6 above)
Comment (bechtol) [reply tests/test_phot_util.py:102]: This is an example of a test where we can create synthetic data. The input is the following array, `mag = np.linspace(20, 25, 101)`, i.e., with intentionally large distribution for testing purposes.

File: tests/data/AB1_design_expected_0_70_i.yaml (hunk 1 above)
Comment (bechtol) [reply tests/data/AB1_design_expected_0_70_i.yaml:10]: Yes, this is the current output of faro. Sounds like it should be flagged and addressed in a separate PR.

**Pull request merged**